### PR TITLE
Combine journal catalog entries with the same id

### PR DESCRIPTION
**Pull request by stefwalter**
Instead of discarding duplicate catalog entries, we now combine them. This allows software or admins to add or override catalog headers, or add additional text to the catalog message.

This was discussed and implemented at systemd.conf Hackfest. I just forgot to make a pull request :S

### CODE REVIEW

**Comment by zonque** [comment]
Before someone gets around to really review this, I take the liberty to comment on some coding style issues first :)

File: src/journal/catalog.c [hunk 1]
```diff
@@ -94,25 +94,86 @@ const struct hash_ops catalog_hash_ops = {
         .compare = catalog_compare_func
 };
 
+static bool next_header(const char **s) {
+        const char *e;
+
+        e = strchr (*s, '\n');
```
Comment (zonque) [inline src/journal/catalog.c]: Please remove white spaces before opening parentheses.

File: src/journal/catalog.c [hunk 2]
```diff
@@ -94,25 +94,86 @@ const struct hash_ops catalog_hash_ops = {
...
+
+        /* End of headers */
+        if (e == *s)
+                return false;
+
+        *s = e + 1;
+        return true;
+}
+
+static char *combine_entries(const char *one, const char *two) {
```
Comment (zonque) [inline src/journal/catalog.c:120]: We generally prefer to return real errors, and pass back allocated pointers through a `**` pointer. This allows proper error propagation up the call stack.

File: src/journal/catalog.c [hunk 3]
```diff
@@ -94,25 +94,86 @@ const struct hash_ops catalog_hash_ops = {
...
+        b1 = one;
+        while (next_header (&b1));
+        b2 = two;
+        while (next_header (&b2));
+
+        l1 = strlen (one);
+        l2 = strlen (two);
+        dest = malloc (l1 + l2 + 1);
+        if (!dest) {
+                log_oom();
```
Comment (zonque) [inline src/journal/catalog.c:133]: With the above change implemented, this would just become `return log_oom();`

File: src/journal/catalog.c [hunk 4]
```diff
@@ -94,25 +94,86 @@ const struct hash_ops catalog_hash_ops = {
...
+
+        /* Find split point of headers to body */
+        b1 = one;
+        while (next_header (&b1));
+        b2 = two;
+        while (next_header (&b2));
+
+        l1 = strlen (one);
+        l2 = strlen (two);
+        dest = malloc (l1 + l2 + 1);
```
Comment (zonque) [inline src/journal/catalog.c]: Please use `new()` in favor of `malloc()`

File: src/journal/catalog.c (hunk 1 above)
Comment (stefwalter) [inline src/journal/catalog.c]: Done. Only for functions right?

File: src/journal/catalog.c (hunk 2 above)
Comment (stefwalter) [inline src/journal/catalog.c:120]: Done.

File: src/journal/catalog.c (hunk 4 above)
Comment (stefwalter) [inline src/journal/catalog.c]: New takes a struct type. Here we're allocating a specific number of bytes.

File: src/journal/catalog.c (hunk 3 above)
Comment (stefwalter) [inline src/journal/catalog.c:133]: Done.

File: src/journal/catalog.c [hunk 5]
```diff
@@ -94,25 +94,85 @@ const struct hash_ops catalog_hash_ops = {
...
+
+        /* Find split point of headers to body */
+        b1 = one;
+        while (next_header(&b1));
+        b2 = two;
+        while (next_header(&b2));
+
+        l1 = strlen(one);
+        l2 = strlen(two);
+        dest = malloc(l1 + l2 + 1);
```
Comment (zonque) [inline src/journal/catalog.c]: We still prefer `new()` over plain `malloc()` for type safety of the returned value.

Comment (stefwalter) [inline src/journal/catalog.c]: Okay. Done.

File: src/journal/catalog.c [hunk 6]
```diff
@@ -94,25 +94,85 @@ const struct hash_ops catalog_hash_ops = {
         .compare = catalog_compare_func
 };
 
+static bool next_header(const char **s) {
```
Comment (vcaputo) [inline src/journal/catalog.c]: This seems mostly duplicitous of find_header(), perhaps get them to reuse this piece?

File: src/journal/catalog.c [hunk 7]
```diff
@@ -94,25 +94,85 @@ const struct hash_ops catalog_hash_ops = {
...
+        return 0;
+}
+
 static int finish_item(
                 Hashmap *h,
-                struct strbuf *sb,
                 sd_id128_t id,
                 const char *language,
-                const char *payload) {
+                char *payload) {
```
Comment (vcaputo) [inline src/journal/catalog.c:169]: at a glance I don't see why const was discarded on payload

File: src/journal/catalog.c [hunk 8]
```diff
@@ -94,25 +94,85 @@ const struct hash_ops catalog_hash_ops = {
         .compare = catalog_compare_func
 };
 
+static bool next_header(const char **s) {
+        const char *e;
+
```
Comment (vcaputo) [inline src/journal/catalog.c:99]: assert(s)

File: src/journal/catalog.c (hunk 7 above)
Comment (stefwalter) [inline src/journal/catalog.c:169]: This function takes ownership of payload.

File: src/journal/catalog.c (hunk 8 above)
Comment (stefwalter) [inline src/journal/catalog.c:99]: Added.

File: src/journal/catalog.c (hunk 7 above)
Comment (stefwalter) [inline src/journal/catalog.c:169]: Added comment to this effect.

Comment (vcaputo) [inline src/journal/catalog.c:169]: Oh, duh,yeah the strbuf used to be getting copied to. It's clear, I just hadn't looked closely enough.

**Comment by poettering** [comment]
OK, looks pretty good to me, but please fix the nitpicks I found. Looks good to merge then!

(And sorry for taking so long to review this!)

**Comment by stefwalter** [comment]
Fixed for review.

**Comment by stefwalter** [comment]
Using streq instead of strcmp, removed further space after function name.

**Pull request merged**